doozerlib/dblib.py

```diff
@@ -21,44 +148,128 @@ def __init__(self, runtime, environment, dry_run=False):
+
+        self.column_identifier = self.create_column_identifier()
+
+    @staticmethod
+    def create_column_identifier():
+        int_handler = IntegerIdentifier()
+        float_handler = FloatIdentifier()
+        datetime_handler = DateTimeIdentifier()
+        varchar_handler = VarcharIdentifier()
+
+        int_handler.set_next_handler(float_handler).set_next_handler(datetime_handler).set_next_handler(varchar_handler)
+        return int_handler
+
+    def check_table_exist(self, table_name):
+        cursor = self.connection.cursor()
+        cursor.execute("""
+            SELECT COUNT(*)
+            FROM information_schema.tables
+            WHERE table_name = '{0}'
+            """.format(table_name.replace('\'', '\'\'')))
+        if cursor.fetchone()[0] == 1:
+            cursor.close()
+            return True
+
+        cursor.close()
+        return False
+
+    def get_missing_columns(self, attr_payload, table_name):
+
+        cursor = self.connection.cursor()
+        cursor.execute(f"show columns from {table_name}")
+        existing_columns = set()
+        payload_columns = set()
+
+        for data in cursor:
+            existing_columns.add(data[0])
+
+        for column in attr_payload:
+            payload_columns.add(column)
+
+        missing_columns = list(payload_columns.difference(existing_columns))
+        return_columns = []
+        for column in missing_columns:
+            return_columns.append((column, self.column_identifier.identify(attr_payload[column])))
+        return return_columns
+
+    def handle_missing_table(self, table_name):
+
+        table_status = self.check_table_exist(table_name)
+
+        if not table_status:
+            cursor = self.connection.cursor()
+            cursor.execute(f'create table {table_name}(log_{table_name}_id bigint auto_increment primary key)')
+            cursor.close()
+
+    def handle_missing_columns(self, payload, table_name):
+
+        missing_columns = self.get_missing_columns(payload, table_name)
+
+        if missing_columns:
+            cursor = self.connection.cursor()
+
+            for column in missing_columns:
+                missing_column = column[0]
+                column_type = column[1][0]
+                cursor.execute(f"alter table {table_name} add column `{missing_column}` {column_type}")
+
+            cursor.close()
+
+    def insert_payload(self, payload, table_name):
+
+        columns, values = [], []
+
+        for column in payload:
+            columns.append(f'`{column}`')
+            values.append(f'"{payload[column]}"')
+
+        columns = ",".join(columns)
+        values = ",".join(values)
+
+        insert_string = f"insert into {table_name}({columns}) values({values})"
+
+        cursor = self.connection.cursor()
+        cursor.execute(insert_string)
+        self.connection.commit()
+        cursor.close()
+
+    @try_connecting
+    def create_payload_entry(self, payload, table_name):
```

Comment:
does this thread? we're definitely running builds in different threads, i would think we write db entries from threads as well...

Reply:
nevermind, i don't think dblib was originally designed to run threaded.

Reply:
Interacting with records should be threadsafe. https://github.com/openshift/doozer/blob/2bfe554eb350bd2219ea0258866489610911fb54/doozerlib/dblib.py#L165 provided the global lock for writing to simpledb. Ultimately, I think these methods should be `_create_payload_entry()` etc so that they are only invoked via Record.

Reply:
@jupierce I am  calling
 `db.create_payload_entry()` 
inside the
 `with self.db.runtime.get_named_semaphore('package::boto3'): `

Shouldn't that make all the inner function calls thread safe by default?

Reply:
It does. The suggested encapsulation just ensures that no one is tempted to call it directly in the future.